hydroflow_lang/src/graph/ops/zip.rs

```diff
@@ -0,0 +1,123 @@
+use quote::quote_spanned;
+use syn::parse_quote;
+
+use super::{
+    FlowProperties, FlowPropertyVal, OpInstGenerics, OperatorConstraints, OperatorInstance,
+    OperatorWriteOutput, Persistence, WriteContextArgs, RANGE_0, RANGE_1,
+};
+use crate::diagnostic::{Diagnostic, Level};
+
+/// > 2 input streams of type `V1` and `V2`, 1 output stream of type `(V1, V2)`
+///
+/// Zips the streams together, forming paired tuples of the inputs. Note that zipping is done
+/// per-tick. Excess items from one input or the other will be discarded. If you do not want to
+/// discard the excess, use [`zip_longest`](#zip_longest) instead.
+///
+/// ```hydroflow
+/// source_iter(0..3) -> [0]my_zip;
+/// source_iter(0..5) -> [1]my_zip;
+/// my_zip = zip() -> for_each(|(l, r)| println!("({}, {})", l, r));
+/// // prints `(0, 0)`, `(1, 1)`, `(2, 2)`.
+/// ```
+#[hydroflow_internalmacro::operator_docgen]
+pub const ZIP: OperatorConstraints = OperatorConstraints {
+    name: "zip",
+    hard_range_inn: &(2..=2),
+    soft_range_inn: &(2..=2),
+    hard_range_out: RANGE_1,
+    soft_range_out: RANGE_1,
+    num_args: 0,
+    persistence_args: &(0..=1),
+    type_args: RANGE_0,
+    is_external_input: false,
+    ports_inn: Some(|| super::PortListSpec::Fixed(parse_quote! { 0, 1 })),
```

Comment:
name them lhs/rhs? Maybe in another PR though, since there are other operators that are like this (like join/crossjoin).

Reply:
Yeah personally using rhs/lhs internally because it is easier, but tuples use tup.0 and tup.1 so matching the ports up with that.

Guessing right-to-left text doesn't really come into play